### PR TITLE
feat(drawbridge): add client secrets login

### PR DESCRIPTION
**Pull request by puiterwijk**
This adds logging in with client credential grant for machine-2-machine interactions, like bots for uploading data to a Drawbridge instance.

Signed-off-by: Patrick Uiterwijk <patrick@profian.com>

### CODE REVIEW

File: src/cli/user/login.rs [hunk 1]
```diff
@@ -14,6 +14,8 @@ pub struct Options {
     oidc_domain: Url,
     #[clap(long, default_value = "4NuaJxkQv8EZBeJKE56R57gKJbxrTLG2")]
     oidc_client_id: String,
+    #[clap(long)]
+    oidc_client_secret: Option<String>,
```
Comment (rvolosatovs) [inline src/cli/user/login.rs:18]: Do we really want to expose this on the command-line?
Especially if this is to be used by CI, using an env variable for this is probably the lesser of all evils

File: src/drawbridge.rs [hunk 2]
```diff
@@ -193,49 +195,53 @@ pub fn login(
...
-    .set_device_authorization_url(dev_auth_url);
+    .set_device_authorization_url(dev_auth_url))
+}
 
-    let details: StandardDeviceAuthorizationResponse = client
-        .exchange_device_code()
-        .context("Failed to construct device authorization request")?
-        .add_scope(Scope::new("openid".into()))
-        .add_scope(Scope::new("profile".into()))
+pub fn login_client_credentials(
```
Comment (rvolosatovs) [inline src/drawbridge.rs:208]: I think there should only be one `login` function, which should simply be extended with additional parameter (`oidc_client_secret`)

That would reduce duplication and simplify maintenance and would prevent errors, like the one made in this PR already missing e.g. this case https://github.com/enarx/enarx/blob/09bce3fd923bbbb52aac185c5858e377b949f043/src/cli/user/register.rs#L56-L61

Since the list of parameters is getting long, we probably want an `AuthContext`, `OidcContext` or something of that sort be taken as an agument

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

File: src/drawbridge.rs [hunk 3]
```diff
@@ -267,11 +273,54 @@ pub fn login(
             .set_password(secret)
             .context("Failed to save user credentials")?;
     }
+
```
Comment (rvolosatovs) [inline src/drawbridge.rs:276]: Please avoid making unrelated diffs like this

File: src/drawbridge.rs (hunk 2 above)
Comment (puiterwijk) [reply src/drawbridge.rs:208]: Well, the thing is that there's basically no overlap between this and the current `login` function: the only line of code they share is the `let client = get_oauth2_client(...)`.
So then I'd basically turn both `login` (maybe rename to `login_device_flow`) and `login_client_credentials` into hidden functions, and just have the public `login` function call either of those.
Is that what you'd prefer?

File: src/drawbridge.rs (hunk 3 above)
Comment (puiterwijk) [reply src/drawbridge.rs:276]: You mean the extra blank line, or the refactoring to a generalized function?

File: src/cli/user/login.rs (hunk 1 above)
Comment (puiterwijk) [reply src/cli/user/login.rs:18]: I personally think having the behaviour of a command radically change to where it does something pretty different (exchange client credentials for access token, instead of print a device key, waiting for that, and then exchanging that for an access token) because of the existence (or non-existence) of a "magical" environment variable is kinda irky.
At least as a CLI flag, it's very explicit.

But if you really prefer an env var, sure.

Comment (rvolosatovs) [reply src/cli/user/login.rs:18]: We should not encourage users to pass secrets on the command line.
This secret belongs in a config file (refs #2021 I suppose).
In CI, writing a config file is cumbersome and an env var is far more common and simpler to use.

Regarding the different behavior - I agree, so it seems like we need another CLI flag, which would enable `m2m` login. It should probably be an enum and default to something like `--login device`, `--login client` being the other option for now.

File: src/drawbridge.rs (hunk 2 above)
Comment (rvolosatovs) [reply src/drawbridge.rs:208]: > Well, the thing is that there's basically no overlap between this and the current `login` function: the only line of code they share is the `let client = get_oauth2_client(...)`.

That's not true, there's also
```
        .add_scope(Scope::new("manage:drawbridge_users".into()))
        .add_scope(Scope::new("manage:drawbridge_repositories".into()))
        .add_scope(Scope::new("manage:drawbridge_tags".into()))
        .add_extra_param("audience", format!("https://{}/", host))
```
block, for example, which I am much more concerned about, since it's much more likely to get out-of-sync as time goes on and result in a painful-to-debug bugs

Additionally, the `login` and `login_client_credentials` function distinction does not make sense to me.

As a follow-up on https://github.com/enarx/enarx/pull/2354/files#r1026469814
Let's add an enum:
```
enum AuthContext{
   Device{ // fields relevant to device flow },
   Client{ // fields relevant to client cred flow },
}
```
And pass that to `login`

Login can then match on the enum and act accordingly. There's no need to split it into further helper functions

Comment (puiterwijk) [reply src/drawbridge.rs:208]: So, the problem with merging the scopes and params between the two is that both methods (`oauth2::Client::exchange_client_credentials` and `oauth2::Client::exchange_device_code`) return totally different structs (`oauth2::ClientCredentialsTokenRequest` and `oauth2::DeviceAuthorizationRequest` respectively), which happen to share those methods. So it's not really possible to extract that into common code.
However, I can use the `add_scopes` for both, and just use a statically set list of scopes, which at least forces those two to be in sync.

Comment (puiterwijk) [reply src/drawbridge.rs]: I made that change right now.
